sycl/plugins/esimd_cpu/pi_esimd_cpu.cpp

```diff
@@ -462,30 +630,104 @@ template <class KernelType, class KernelArgType, int DIMS> class libCMBatch {
     GroupDim[1] = (uint32_t)(GlobalSize[1] / LocalSize[1]);
     GroupDim[2] = (uint32_t)(GlobalSize[2] / LocalSize[2]);
 
-    PLACEHOLDER_UNUSED(WrappedLambda_NDITEM_3DIM);
-    // TODO : Invoke invokeLambda_NDITEM_3DIM through CM's multi-threaded
-    // kernel launching with WrappedLambda_NDITEM_3DIM and dimension info
+    ESimdCPUKernel ESimdCPU((fptrVoid)InvokeLambda_NDITEM_3DIM, GroupDim,
+                            SpaceDim);
+
+    ESimdCPU.launchMT(sizeof(struct LambdaWrapper_NDITEM_3DIM),
+                      WrappedLambda_NDITEM_3DIM.get());
   }
 };
 
+/// Implementation for ESIMD_CPU device interface accessing ESIMD
+/// intrinsics and LibCM functionalties requred by intrinsics
 // Intrinsics
 sycl::detail::ESIMDDeviceInterface::ESIMDDeviceInterface() {
-  reserved = nullptr;
   version = ESIMDEmuPluginInterfaceVersion;
+  reserved = nullptr;
 
-  /// TODO : Fill *_ptr fields with function pointers from CM
-  /// functions prefixed with 'cm_support'
+  /* From 'esimd_emu_functions_v1.h' : Start */
+  cm_barrier_ptr = cm_support::barrier;
+  cm_sbarrier_ptr = cm_support::split_barrier;
+  cm_fence_ptr = cm_support::fence;
 
-  cm_barrier_ptr = nullptr;  /* cm_support::barrier; */
-  cm_sbarrier_ptr = nullptr; /* cm_support::split_barrier; */
-  cm_fence_ptr = nullptr;    /* cm_support::fence; */
+  sycl_get_surface_base_addr_ptr = cm_support::get_surface_base_addr;
+  __cm_emu_get_slm_ptr = cm_support::get_slm_base;
+  cm_slm_init_ptr = cm_support::init_slm;
+  /* From 'esimd_emu_functions_v1.h' : End */
+}
+
+/// Implementation for Host Kernel Launch used by
+/// piEnqueueKernelLaunch
+template <typename T> using KernelFunc = std::function<void(T)>;
 
-  sycl_get_surface_base_addr_ptr =
-      nullptr;                    /* cm_support::get_surface_base_addr; */
-  __cm_emu_get_slm_ptr = nullptr; /* cm_support::get_slm_base; */
-  cm_slm_init_ptr = nullptr;      /* cm_support::init_slm; */
+template <int NDims> struct InvokeBaseImpl {
+  static sycl::range<NDims> get_range(const size_t *GlobalWorkSize);
+};
+
+static constexpr bool isNull(int NDims, const size_t *R) {
+  return ((0 == R[0]) && (1 > NDims || 0 == R[1]) && (2 > NDims || 0 == R[2]));
 }
 
+template <int NDims, typename ArgTy> struct InvokeImpl {
```

Comment:
need comments, plus describe what is NDims and ArgTy

Reply:
NDims is the number of dimensions in the ND-range. ArgTy is nd_item<NDims>, the argument type of normalized kernel. Kernels are normalized in the handler so that all kernels take an nd_item as argument (see StoreLambda in CL/sycl/handler.hpp). For kernels whose workgroup size (LocalWorkSize) is unspecified, InvokeImpl sets LocalWorkSize to {1, 1, 1}, i.e. each workgroup contains just one work item. CM emulator will run several workgroups in parallel depending on environment settings.

Reply:
Dasarath's comment is inserted.